### PR TITLE
mount unit state engine fixes

### PR DESCRIPTION
**Pull request by poettering**
A number of fixes in order to get #6048 settled. Given that these changes and their impact aren't obvious, this is really stuff to eye the CI state for...

### CODE REVIEW

File: src/basic/set.h [hunk 1]
```diff
@@ -136,3 +136,5 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(Set*, set_free_free);
 
 #define _cleanup_set_free_ _cleanup_(set_freep)
 #define _cleanup_set_free_free_ _cleanup_(set_free_freep)
+
+int set_make(Set **ret, const struct hash_ops *hash_ops HASHMAP_DEBUG_PARAMS, void *add, ...);
```
Comment (keszybz) [inline src/basic/set.h:140]: No tests? :(

File: src/core/service.c [hunk 2]
```diff
@@ -1782,14 +1781,34 @@ static void service_enter_start_post(Service *s) {
...
+        int r;
 
-        if (!UNIT(s)->cgroup_path)
-                return;
+        assert(s);
+
+        if (s->control_pid > 0) {
+                r = kill_and_sigcont(s->control_pid, SIGKILL);
+                if (r < 0)
+                        log_unit_debug_errno(UNIT(s), r, "Failed to kill control process, ignoring: %m");
```
Comment (keszybz) [inline src/core/service.c]: It'd be nice to print the PID here (and possibly also comm, but maybe that's too much). Just to make any debugging easier (and NFS-related process might be unkillable, so this is not entirely theoretical). "SIGKILL" should be mentioned too.

File: src/core/service.c [hunk 3]
```diff
@@ -1782,14 +1781,34 @@ static void service_enter_start_post(Service *s) {
...
+                if (s->control_pid > 0) {
+                        r = set_make(&pid_set, PID_TO_PTR(s->control_pid), NULL);
+                        if (r < 0)
+                                return log_oom();
+                }
+
+                p = strjoina(UNIT(s)->cgroup_path, "/control");
+                r = cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, p, SIGKILL, CGROUP_SIGCONT|CGROUP_IGNORE_SELF|CGROUP_REMOVE, pid_set, NULL, NULL);
+                if (r < 0)
+                        log_unit_debug_errno(UNIT(s), r, "Failed to kill control process cgroup: %m");
```
Comment (keszybz) [inline src/core/service.c]: `p` and `SIGKILL` should appear in the message.

File: src/core/service.c [hunk 4]
```diff
@@ -1782,14 +1781,34 @@ static void service_enter_start_post(Service *s) {
         service_enter_stop(s, SERVICE_FAILURE_RESOURCES);
 }
 
-static void service_kill_control_processes(Service *s) {
-        char *p;
+static int service_kill_control_processes(Service *s) {
+        int r;
```
Comment (keszybz) [inline src/core/service.c]: The return value is not checked anywhere...

File: src/core/mount.c [hunk 5]
```diff
@@ -1322,72 +1329,53 @@ static int mount_dispatch_timer(sd_event_source *source, usec_t usec, void *user
...
-                        log_unit_warning(UNIT(m), "Remounting timed out. Skipping SIGKILL. Ignoring.");
+        case MOUNT_REMOUNTING_SIGKILL:
+                mount_set_reload_result(m, MOUNT_FAILURE_TIMEOUT);
 
-                        if (m->from_proc_self_mountinfo)
-                                mount_enter_mounted(m, MOUNT_FAILURE_TIMEOUT);
-                        else
-                                mount_enter_dead(m, MOUNT_FAILURE_TIMEOUT);
-                }
+                log_unit_warning(UNIT(m),"Mount process still around after SIGKILL. Ignoring.");
```
Comment (keszybz) [inline src/core/mount.c]: There's a bunch of places where a space is missing after `UNIT(m),`.

Would be nice to include the PID (and comm?) of the stubborn process.

**Comment by keszybz** [comment]
LGTM, but the logic is pretty complicated. It behooves us to test this with hanging NFS mounts.

@neilbrown any chance you could test this?

**Comment by keszybz** [comment]
I'm setting the r/n-r flag because at least the unchecked return value needs some changes.

File: src/basic/set.h (hunk 1 above)
Comment (poettering) [reply src/basic/set.h:140]: will add

File: src/core/mount.c (hunk 5 above)
Comment (poettering) [reply src/core/mount.c]: unit_kill_context() actually prints that already, which is called by mount_enter_signal(), hence we should be good in this regard already.

**Comment by poettering** [comment]
Force pushed a new version. No changes except for the suggested ones (but I don#t do the second kill logging thing for the reason pointed out). Please have another look!

Thanks!

**Comment by neilbrown** [comment]
Thanks!  I briefly tested with these patches and they appear to make systemd work exactly as I think it should in this circumstance.

**Pull request merged**